### PR TITLE
feat(api): Validate plate reader status using live data hookups for engine and introduce lid status to the PAPI

### PR DESCRIPTION
**Pull request by CaseyBatten**
# Overview
Ensure that the lid position is automatically updated in live runs, and that the live data is used to validate position on load, open and close. Validate that labware cannot be moved to the plate reader unless the plate reader is open. Add `is_lid_on()` PAPI exposure for use in python protocols.

Closes: PLAT-208 PLAT-386

## Test Plan and Hands on Testing
Run the following protocol and ensure the expected results below occur (NOTE: Must set Max API version to 2.21):
```
from opentrons import protocol_api

# metadata
metadata = {
    'protocolName': 'Absorbance Reader Test Protocol',
    'author': 'Platform Expansion',
}

requirements = {
    "robotType": "Flex",
    "apiLevel": "2.21",
}

DRYRUN = 'YES'
USE_GRIPPER = 'TRUE'

# protocol run function
def run(protocol: protocol_api.ProtocolContext):

    mod = protocol.load_module('absorbanceReaderV1', 'D3')
    sample_plate = protocol.load_labware('armadillo_96_wellplate_200ul_pcr_full_skirt','A2')

    #COMMENT OUT THE FOLLOWING TWO LINES FOR TEST 3
    if mod.is_lid_on():
        mod.open_lid()

    protocol.move_labware(
        sample_plate,
        new_location=mod,
        use_gripper=USE_GRIPPER,
    )
    mod.close_lid()
```
Tests:
- [x] Begin Protocol on Robot with lid on, ensure lid status check passes and lid removal, labware movement, and lid replacement occur as expected
- [x] Begin Protocol on Robot with lid off, ensure .open(...) command no ops and moves on
- [x] Begin Protocol with Lid off, ensure .close(...) command no ops and moves on

## Changelog

- Updated protocol_api/protocol_context.py `move_labware(...)` function to contain special case validation for moving to a plate reader module when it's lid status is closed (expect error to raise). 
- Added in live data hookup infrastructure for the Absorbance Reader, hooked lid status data into the engine to report live (remaining live data points will be passed into engine in small follow up PR)
- Addition of Protocol API `is_lid_on()` method

## Review requests
Take a look at the validation workaround that had to be put into `move_labware(...)`. It's a special casing due to the way that plate reader lid labware is being injected into the engine labware store via the deck configuration. Previously, if the lid was closed and an attempt was made to move labware onto the plate reader, we would get a random key error with the unique labware ID of the lid returned. This addition returns a much nicer error, but I'm not entirely sure if special casing is the best approach here, let me know any thoughts on this.

## Risk assessment
Low risk - only effects plate reader module data and validations.

### CODE REVIEW

**Review by vegano1 — APPROVED** [review]
This makes sense to me!

File: api/src/opentrons/protocol_api/module_contexts.py [hunk 1]
```diff
@@ -994,9 +994,14 @@ def close_lid(self) -> None:
 
     @requires_version(2, 21)
     def open_lid(self) -> None:
-        """Close the lid of the Absorbance Reader."""
+        """Open the lid of the Absorbance Reader."""
         self._core.open_lid()
 
+    @requires_version(2, 21)
+    def is_lid_on(self) -> bool:
+        """Return True if the Absorbance Reader's lid is currently closed."""
```
Comment (ecormany) [inline api/src/opentrons/protocol_api/module_contexts.py]: Let's Sphinx it up
```suggestion
        """Return ``True`` if the Absorbance Reader's lid is currently closed."""
```
I'm OK with leaving the implication that `False` means it's open, but if false means open or unknown, we should state that.

File: api/src/opentrons/protocol_api/protocol_context.py [hunk 2]
```diff
@@ -707,6 +708,13 @@ def move_labware(
                 f"Expected labware of type 'Labware' but got {type(labware)}."
             )
 
+        # Ensure that when moving to an absorbance reader than the lid is open
+        if isinstance(new_location, AbsorbanceReaderContext):
+            if new_location.is_lid_on():
+                raise CommandPreconditionViolated(
+                    f"Cannot move {labware.name} onto the Absorbance Reader Module when Lid is Closed."
```
Comment (ecormany) [inline api/src/opentrons/protocol_api/protocol_context.py]:
```suggestion
                    f"Cannot move {labware.name} onto the Absorbance Reader Module when its lid is closed."
```
nit

File: api/src/opentrons/protocol_engine/commands/absorbance_reader/close_lid.py [hunk 3]
```diff
@@ -108,6 +125,21 @@ async def execute(
...
         )
+
+        if not self._state_view.config.use_virtual_modules:
+            abs_reader = self._equipment.get_module_hardware_api(mod_substate.module_id)
+
+            if abs_reader is not None:
+                result = await abs_reader.get_current_lid_status()
+                if result is not AbsorbanceReaderLidStatus.ON:
+                    raise CannotPerformModuleAction(
+                        "The Opentrons Plate Reader lid mechanicaly position did not match expected Closed state."
```
Comment (ecormany) [inline api/src/opentrons/protocol_engine/commands/absorbance_reader/close_lid.py]: can this error ever bubble up to be shown to users? if so, needs an edit.
```suggestion
                        "The mechanical position of the Absorbance Plate Reader's lid did not match the expected 'closed' state."
```

**Review by ecormany — COMMENTED** [review]
Suggestions for some potentially user-facing copy.

**Comment by vegano1** [comment]
Looking good to me, i tested this on a PlatypusExpansion robot

**Pull request merged**